### PR TITLE
Write samples JSON file to GCS on indexing

### PR DESCRIPTION
**Pull request by bfcrampton**
This will be used in conjunction with the GCS [compose API](https://cloud.google.com/storage/docs/json_api/v1/objects/compose) to significantly shorten the time to write the JSON for exproting to Saturn.

### CODE REVIEW

File: bigquery/indexer.py [hunk 1]
```diff
@@ -261,6 +292,8 @@ def main():
                     sample_id_column, sample_file_columns)
         index_fields(es, index_name + '_fields', table, sample_id_column)
 
+    create_samples_json_export_file(es, index_name, args.billing_project_id)
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

[the remark this answers is not in the document]
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py [hunk 2]
```diff
@@ -240,6 +242,35 @@ def read_table(client, table_name):
         client.dataset(dataset_id, project=project_id).table(table_name))
 
 
+def create_samples_json_export_file(es, index_name, project_id):
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py [hunk 3]
```diff
@@ -240,13 +243,56 @@ def read_table(client, table_name):
...
+    """
+    Writes the samples export JSON file to a GCS bucket. This significantly 
+    speeds up exporting the samples table to Saturn in the Data Explorer.
+
+    Args:
+        es: Elasticsearch object.
+        index_name: Name of Elasticsearch index.
+        project_id: Google Cloud Project ID containing the export samples bucket
```
Comment (melissachang) [inline bigquery/indexer.py]: project_id: deploy project

File: bigquery/indexer.py [hunk 4]
```diff
@@ -240,13 +243,56 @@ def read_table(client, table_name):
...
+    entities = []
+    search = Search(using=es, index=index_name)
+    for hit in search.scan():
+        participant_id = hit.meta['id']
+        doc = hit.to_dict()
+        for sample in doc.get('samples', []):
+            sample_id = sample['sample_id']
+            export_sample = {'participant': participant_id}
+            for table_column, value in sample.iteritems():
+                # Ignore _has_* and sample_id fields.
```
Comment (melissachang) [inline bigquery/indexer.py]: (It took me a while to understand this)

table_column -> es_field_name
```
                # es_field_name looks like "_has_chr_18_vcf", "sample_id" or "verily-public-data.human_genome_variants.1000_genomes_sample_info.In_Low_Coverage_Pilot".
                splits = table_column.split('.')
                # Ignore _has_* and sample_id fields.
                if len(splits) != 4:
```

File: bigquery/indexer.py [hunk 5]
```diff
@@ -240,13 +243,56 @@ def read_table(client, table_name):
...
+            entities.append({
+                'entityType': 'sample',
+                'name': sample_id,
+                'attributes': export_sample,
+            })
+
+    client = storage.Client(project=project_id)
+    bucket_name = '%s-export-samples' % project_id
+    try:
+        bucket = client.get_bucket(bucket_name)
```
Comment (melissachang) [inline bigquery/indexer.py]: Use [lookup_bucket](https://googleapis.github.io/google-cloud-python/latest/storage/client.html#google.cloud.storage.client.Client.lookup_bucket) instead, then you don't need to catch an exception

File: bigquery/indexer.py [hunk 6]
```diff
@@ -240,13 +243,56 @@ def read_table(client, table_name):
         client.dataset(dataset_id, project=project_id).table(table_name))
 
 
+def create_samples_json_export_file(es, index_name, project_id):
```
Comment (melissachang) [inline bigquery/indexer.py]: Looks like the file has 1 line with 20918223 characters. Can you split into multiple lines, maybe one line per sample?

File: bigquery/indexer.py [hunk 7]
```diff
@@ -240,13 +243,56 @@ def read_table(client, table_name):
...
+                export_sample[splits[3]] = value
+
+            entities.append({
+                'entityType': 'sample',
+                'name': sample_id,
+                'attributes': export_sample,
+            })
+
+    client = storage.Client(project=project_id)
+    bucket_name = '%s-export-samples' % project_id
```
Comment (melissachang) [inline bigquery/indexer.py]:
```
# Don't put in project_id-export because that bucket has TTL= 1 day.
```

Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py [hunk 8]
```diff
@@ -240,13 +243,56 @@ def read_table(client, table_name):
         client.dataset(dataset_id, project=project_id).table(table_name))
 
 
+def create_samples_json_export_file(es, index_name, project_id):
+    """
+    Writes the samples export JSON file to a GCS bucket. This significantly 
+    speeds up exporting the samples table to Saturn in the Data Explorer.
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py (hunk 3 above)
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py (hunk 4 above)
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py (hunk 5 above)
Comment (bfcrampton) [reply bigquery/indexer.py]: Done.

File: bigquery/indexer.py (hunk 6 above)
Comment (bfcrampton) [reply bigquery/indexer.py]: Done--formatted `json.dumps()`.

**Pull request merged**